prow/plugins/config_test.go

```diff
@@ -1520,3 +1523,141 @@ orgA/repoB:
 		t.Error("expected unmarshal error but didn't get one")
 	}
 }
+
+func TestConfigMergingProperties(t *testing.T) {
+	t.Parallel()
+	testCases := []struct {
+		name          string
+		makeMergeable func(*Configuration)
+	}{
+		{
+			name: "Plugins config",
+			makeMergeable: func(c *Configuration) {
+				*c = Configuration{Plugins: c.Plugins}
+			},
+		},
+	}
+
+	expectedProperties := []struct {
+		name         string
+		verification func(t *testing.T, fuzzedConfig *Configuration)
+	}{
+		{
+			name: "Merging into empty config always succeeds and makes the empty config equal to the one that was merged in",
+			verification: func(t *testing.T, fuzzedMergeableConfig *Configuration) {
+				newConfig := &Configuration{}
+				if err := newConfig.mergeFrom(fuzzedMergeableConfig); err != nil {
+					t.Fatalf("merging fuzzed mergeable config into empty config failed: %v", err)
+				}
+				if diff := cmp.Diff(newConfig, fuzzedMergeableConfig); diff != "" {
+					t.Errorf("after merging config into an empty config, the config that was merged into differs from the one we merged from:\n%s\n", diff)
+				}
+			},
+		},
+		{
+			name: "Merging empty config in always succeeds",
+			verification: func(t *testing.T, fuzzedMergeableConfig *Configuration) {
+				if err := fuzzedMergeableConfig.mergeFrom(&Configuration{}); err != nil {
+					t.Errorf("merging empty config in failed: %v", err)
+				}
+			},
+		},
+		{
+			name: "Merging a config into itself always fails",
+			verification: func(t *testing.T, fuzzedMergeableConfig *Configuration) {
+				if apiequality.Semantic.DeepEqual(fuzzedMergeableConfig, &Configuration{}) {
+					return
+				}
+
+				if err := fuzzedMergeableConfig.mergeFrom(fuzzedMergeableConfig); err == nil {
+					serialized, serializeErr := yaml.Marshal(fuzzedMergeableConfig)
+					if serializeErr != nil {
+						t.Fatalf("merging non-empty config into itself did not yield an error and serializing it afterwards failed: %v. Raw object: %+v", serializeErr, fuzzedMergeableConfig)
+					}
+					t.Errorf("merging a config into itself did not produce an error. Serialized config:\n%s", string(serialized))
+				}
+			},
+		},
+	}
+
+	seed := time.Now().UnixNano()
+	// Print the seed so failures can easily be reproduced
+	t.Logf("Seed: %d", seed)
+	fuzzer := fuzz.NewWithSeed(seed)
```

Comment:
The testing is very reliable actually, we only had one case where it was flaky in the past and that was due to the fuzzer only creating empty, non-nil structs with a very low probability: https://github.com/kubernetes/test-infra/issues/21507

It is IMHO important to have it in a presubmit, because there a bunch of things you simply can not test without, like the "merging doesn't lose data". There is no other test you could write that would start failing if someone adds another field to a struct that is being merged.

FWIW we use fuzz testing in t-i since November already, it was first added in 7fe0b85d44a

We can ofc reconsider this if we see that these tests get unreliable/flaky, but that wasn't the case so far.

Reply:
If it is passing it isn't doing anything useful.

If it fails someone is going to do `/retest` and then because of the randomness it will pass.

So I'm not really seeing the scenario where this is a useful presubmit.

> There is no other test you could write that would start failing if someone adds another field to a struct that is being merged.

Reflection?

Reply:
> If it is passing it isn't doing anything useful.

It won't pass, it will instantly start failing. For a simple example, do a change like this:
```
diff --git a/prow/apis/prowjobs/v1/types.go b/prow/apis/prowjobs/v1/types.go
index 5fabefdbb7..b62ff84c35 100644
--- a/prow/apis/prowjobs/v1/types.go
+++ b/prow/apis/prowjobs/v1/types.go
@@ -325,6 +325,7 @@ func (d *Duration) MarshalJSON() ([]byte, error) {
 // This is primarily used to provide automatic integration with gubernator
 // and testgrid.
 type DecorationConfig struct {
+       TestField string
        // Timeout is how long the pod utilities will wait
        // before aborting a job with SIGINT.
        Timeout *Duration `json:"timeout,omitempty"
```

Then run
```
go test ./prow/apis/prowjobs/v1 -run TestApplyDefaultsAppliesDefaultsForAllFields
```

I'll buy you a beer if you manage to get that test to pass just a single time.

> Reflection?

That is this lib, we'd just be rebuilding it.

Reply:
oh no, I've been nerd sniped!

Reply:
I am defeated:
```
//prow/apis/prowjobs/v1:go_default_test                                  FAILED in 5000 out of 5000 in 1.4s
  Stats over 5000 runs: max = 1.4s, min = 0.3s, avg = 0.6s, dev = 0.2s
```